### PR TITLE
fix(src): drop unused revisions

### PR DESCRIPTION
**Pull request by alarso16**
As described in #865, even after a consumer drops an `Arc` to a proposal, the `RevisionManager` maintains a clone of all uncommitted proposals, which will cause the memory throughout runtime to grow, even though no users have access to the data anymore. A test is added to check that once a user-side `Arc<Proposal<'_>>` is dropped, and another proposal is committed, it will be removed from the list, and the `NodeStore` is no longer accessible. This is implemented by checking the number of strong references on commit for all proposals.

### CODE REVIEW

File: firewood/src/manager.rs [hunk 1]
```diff
@@ -191,7 +191,9 @@ impl RevisionManager {
 
         // 8. Proposal Cleanup
         // first remove the committing proposal from the list of outstanding proposals
-        self.proposals.retain(|p| !Arc::ptr_eq(&proposal, p));
+        // Any proposals that are only referenced in this list will be freed
+        self.proposals
+            .retain(|p| !Arc::ptr_eq(&proposal, p) && Arc::strong_count(p) > 1);
```
Comment (rkuris) [inline firewood/src/manager.rs:196]: The comment here could be cleaner:
```
// Free proposal that is being committed as well as any proposals no longer
// referenced by anyone else

**Review by rkuris — CHANGES_REQUESTED** [review]
Comments mostly related to testing the functionality outside of the API instead of peeking under the covers at the implementation.

This needs a test for R1->P1->P2->P3 as well, since that edge case is possibly a bigger problem. If P2 is dropped, there should still be a reference to it in P3, so we should check that the data in P2 is still accessible through P3.

File: firewood/src/db.rs [hunk 2]
```diff
@@ -523,6 +525,117 @@ mod test {
...
+        // the second proposal shouldn't be available to commit anymore
+        assert!(!db.all_hashes().await.unwrap().contains(&p2hash));
+
+        // the third proposal should still be contained within the all_hashes list
+        // would be deleted if another proposal was committed and proposal3 was dropped here
+        let hash3 = proposal3
+            .nodestore
+            .root_hash()
+            .unwrap()
+            .expect("should have a root hash");
```
Comment (rkuris) [inline firewood/src/db.rs:579]: nit: I usually just use expect in production code, not in tests. If the test fails because the proposal3 doesn't have a root hash, we have huge problems anyway :)

File: firewood/src/db.rs [hunk 3]
```diff
@@ -523,6 +525,117 @@ mod test {
...
+            value: b"v1",
+        }];
+        let proposal1 = db.propose(batch1).await.unwrap();
+        assert_eq!(&*proposal1.val(b"k1").await.unwrap().unwrap(), b"v1");
+
+        let batch2 = vec![BatchOp::Put {
+            key: b"k2",
+            value: b"v2",
+        }];
+        let proposal2 = proposal1.clone().propose(batch2).await.unwrap();
```
Comment (rkuris) [inline firewood/src/db.rs:602]: nit: not sure why the clone was needed here, but it's a test and it doesn't hurt anything.

File: firewood/src/db.rs [hunk 4]
```diff
@@ -523,6 +525,117 @@ mod test {
...
+        let committed = db.root_hash().await.unwrap().unwrap();
+        let historical = db.revision(committed).await.unwrap();
+        assert_eq!(&*historical.val(b"k1").await.unwrap().unwrap(), b"v1");
+
+        // the second proposal shouldn't be available to commit anymore
+        assert!(!db.all_hashes().await.unwrap().contains(&p2hash));
+
+        // the third proposal should still be contained within the all_hashes list
+        let hash3 = proposal3.root_hash().await.unwrap().unwrap();
+        assert!(db.manager.read().unwrap().all_hashes().contains(&hash3));
```
Comment (rkuris) [inline firewood/src/db.rs:631]:
```suggestion
        assert!(db.all_hashes().await.unwrap().contains(&hash3));
```

**Review by rkuris — CHANGES_REQUESTED** [review]
